language-server/test/dotty/tools/languageserver/util/CodeTester.scala

```diff
@@ -0,0 +1,160 @@
+package dotty.tools.languageserver.util
+
+import dotty.tools.languageserver.util.Code.SourceWithPositions
+import dotty.tools.languageserver.util.actions._
+import dotty.tools.languageserver.util.embedded.CodeMarker
+import dotty.tools.languageserver.util.server.{TestFile, TestServer}
+import org.eclipse.lsp4j.{CompletionItemKind, DocumentHighlightKind}
+
+/**
+ * Simulates an LSP client for test in a workspace defined by `sources`.
+ *
+ * @param sources The list of sources in the workspace
+ * @param actions Unused
+ */
+class CodeTester(sources: List[SourceWithPositions], actions: List[Action]) {
+
+  private val testServer = new TestServer(TestFile.testDir)
+
+  private val files = sources.zipWithIndex.map { case (code, i) =>
+    testServer.openCode(code.text, s"Source$i.scala")
+  }
+  private val positions: PositionContext = getPositions(files)
+
+  /**
+   * Perform a hover over `range`, verifies that result matches `expected`.
+   *
+   * @param range    The range over which to hover.
+   * @param expected The expected result.
+   * @return This `CodeTester` after performing the action.
+   *
+   * @see dotty.tools.languageserver.util.actions.CodeHover
+   */
+  def hover(range: CodeRange, expected: String): CodeTester =
```

Comment:
This method and the following ones could return `this.type` instead of `CodeTester`, that way you don't have to document `@return`, it's obvious.